docs/syntax_grammar.md

```diff
@@ -0,0 +1,427 @@
+
+    type       = BASIC_TYPE | array_type | map_type | "any" .
+    BASIC_TYPE = "num" | "string" | "bool" .
+    array_type = type "[]" .
+    map_type   = [type] "{}" .
+
+    /* --- Expressions --- */
+    toplevel_expr = func_call | type_assertion | expr.
+
+    func_call = ident args .
+    args      = { unary_expression } .
+
+    type_assertion = type ( ident | ident_selector ) .
+    ident_selector = ident selector { selector } .
+    selector       = index | slice | dot_selector .
+    index          = "[" expr "]" .
+    slice          = "[" [expr] : [expr] "]" .
+    dot_selector   = "." ident .
+
+    expr       = unary_expr | expr OP expr .
+    unary_expr = operand | UNARY_OP unary_expr .
+    operand    = literal | ident | ident_selector | "(" toplevel_expr ")" .
+    UNARY_OP   = "+" | "-" | "!" .
+    OP         = "or" | "and" | REL_OP | ADD_OP | MUL_OP .
+    REL_OP     = "==" | "!=" | "<" | "<=" | ">" | ">=" .
+    ADD_OP     = "+" | "-" .
+    MUL_OP     = "*" | "/" | "%" .
+
+    /* --- Literals --- */
+    literal     =  num_lit | string_lit | BOOL_CONST | array_lit | map_lit .
+    num_lit     =  DECIMAL_DIGIT { DECIMAL_DIGIT } |
+                   DECIMAL_DIGIT { DECIMAL_DIGIT } "." { DECIMAL_DIGIT } .
+    string_lit  = """ { UNICODE_CHAR } """ .
+    BOOL_CONST  = "true" | "false" .
+    array_lit   = type "[" array_elems "]" .
+    array_elems = { unary_expr [NL] }
+    map_lit     = [type] "{" map_elems "}" .
+    map_elems   = { ident ":" unary_expr [NL] } .
+    
+    /* --- Control flow --- */
+    loop       = for | while .
+    for        = "for" range NL
+                     statements
+                 END .
+    range      = ident ( ":=" | "=" ) "range" range_args .
+    range_args = unary_expr [ unary_expr [ unary_expr ] ] .
+    while      = "while" toplevel_expr NL
+                     statements
+                 END .
+
+    if = "if" toplevel_expr NL
+                statements
+          { "else" "if" toplevel_expr NL
+                statements }
+          [ "else" NL
+                statements ]
+          END .
+
+    /* --- Functions ---- */
+    func            = "func" func_signature NL
+                          statements
+                      END .
+    func_signature  = ident [ ":" type ] params .
+    params          = { param } | variadic_param .
+    param           = ident ":" type .
+    variadic_param  = param "..." .
+    return          = "return" [toplevel_expr] .
+
+    event_handler   = "on" ident NL
+                          statements
+                      END .
+
+    /* --- Terminals --- */
+    LETTER         = UNICODE_LETTER | "_" .
+    UNICODE_LETTER = /* a Unicode code point categorized as "Letter" (category L) */ .
+    UNICODE_DIGIT  = /* a Unicode code point categorized as "Number, decimal digit" */ .
+    UNICODE_CHAR   = /* an arbitrary Unicode code point except newline */ .
+    DECIMAL_DIGIT  = "0" … "9" .
+    END            = "end" .
```

Comment:
I think i would get rid of this and just use the literal `"end"` in its place. I know it appears more times than any other literal, but this is the only word that is defined this way.

Reply:
ok, good idea, done.